File: C++/SortingAlgorithmsComparison/src/benchmarker.cpp
//
// Created by Franc on 30/10/2025.
//

#include "../include/benchmarker.hpp"

BenchmarkResults Benchmarker::foo(const ArrayFunction& algorithm) const
{
    BenchmarkResults results;
    results.reserve(sample_generators.size());
    for (const auto& sample_generator : sample_generators)
    {
        std::vector<PreciseMilliseconds> sample_type_results;
        for (const auto& sample_length : sample_sizes)
        {
            sample_generator(test_data_buffer.get(), sample_length);

            const auto start = std::chrono::steady_clock::now();
            algorithm(test_data_buffer.get(), sample_length);
            const auto end = std::chrono::steady_clock::now();

            const PreciseMilliseconds time_taken = end - start;
            sample_type_results.push_back(time_taken);
        }

        results.push_back(sample_type_results);
    }

    return results;
}

void Benchmarker::addAverages(BenchmarkResults& results) const
{
    const std::size_t types = sample_generators.size();
    const std::size_t sizes = sample_sizes.size();

    // Add sample type average
    for (auto & result : results)
    {
        float type_average = 0;
        for (auto size_index : result)
        {
            type_average += size_index.count();
        }
        type_average /= static_cast<float>(sizes);
        result.emplace_back(type_average);
    }

    // Add sample size average + algorithm average
    std::vector<float> size_averages;
    size_averages.resize(sizes+1);
    for (auto & result : results)
    {
        for (int i = 0; i < sizes+1; i++)
        {
            size_averages[i] += result[i].count() / static_cast<float>(types);
        }
    }

    results.resize(types+1);
    for (auto size_average : size_averages)
    {
        results.back().emplace_back(size_average);
    }
}

BenchmarkResults Benchmarker::runBenchmark(const ArrayFunction& algorithm) const
{
    BenchmarkResults results = foo(algorithm);
    addAverages(results);

    return results;
}

File: JavaScript/Benchmarker+/benchmarker.js
class Benchmarker {
    constructor(sample_sizes, sample_generators) {
        this.sample_sizes = sample_sizes;
        this.sample_generators = sample_generators;
    }

    runBenchmark(sorting_algorithm) {
        let results = this.#getTimes(sorting_algorithm);
        this.#addAverages(results);

        return results;
    }

    #getTimes(sorting_algorithm) {
        let results = new Array(sample_generators.length);
        for (let i = 0; i < sample_generators.length; i++) {
            results[i] = new Array(samples_size.length).fill(-1);
        }

        let i = 0;
        for (const generator of sample_generators) {
            let j = 0;
            for (const size of samples_size) {
                let sample = new Array(size);
                generator(sample);
                // if (size === 100) {
                //     console.log(generator.name);
                //     console.log(sample);
                // }

                const start = performance.now();
                sorting_algorithm(sample);
                const end = performance.now();

                results[i][j] = end - start;
                j++;
            }
            i++;
        }
        return results;
    }

    #addAverages(results) {
        const types = sample_generators.length;
        const sizes = samples_size.length;

        // Sample type averages
        for (const sample_type_results of results) {
            let type_average = 0;
            for (const size_result of sample_type_results) {
                type_average += Number(size_result);
            }

            sample_type_results.push(type_average / sizes);
        }

        // Size average + total average
        let size_averages = new Array(sizes+1).fill(0);
        for (const sample_type_results of results) {
            for (let i = 0; i <= sizes; i++) {
                size_averages[i] += sample_type_results[i] / types;
            }
        }

        results.push(size_averages);
    }
}